File: src/leetCode/Ex77.java
package leetCode;

import java.util.ArrayList;
import java.util.List;

public class Ex77 {
    private List<List<Integer>> res;
    private List<Integer> temp;
    int len;
    int N;
    public List<List<Integer>> combine(int n, int k) {
        res = new ArrayList<>();
        temp = new ArrayList<>();
        len = k;
        N = n;
        helper(1, 0);
        return res;
    }

    private void helper(int i, int k){
        if(N-i+1<k-temp.size()){
            return;
        }
        
        if(k == len){
            res.add(new ArrayList<>(temp));
            return;
        }
        // i not been chosen
        helper(i+1, k);
        temp.add(i);
        helper(i+1, k+1);
        temp.remove(temp.size()-1);
    }

    public static void main(String[] args) {
        var o = new Ex77();
        var t = o.combine(4, 2);
        for (List<Integer> list : t) {
            System.out.println(list.toString());
        }
    }
}

File: src/leetCode/ex77.kt
package leetCode

fun combine(n: Int, k: Int): List<List<Int>> {
    val res = mutableListOf<MutableList<Int>>()
    combineDfs(mutableListOf(),res,0,k,n,0)
    return res
}

private fun combineDfs(list:MutableList<Int>,result:MutableList<MutableList<Int>>,k:Int,target:Int,n:Int,item:Int){
    if (k==target){
        result.add(mutableListOf<Int>().apply { addAll(list) })
    }
    (item+1 .. n ).forEach {
        list.add(it)
        combineDfs(list,result,k+1,target,n,it)
        list.remove(it)
    }
}

fun main(args: Array<String>) {
    combine(4,2).forEach { println(it) }
}